# Calculate the percentage of calories for carbohydrates, proteins, and fats without using an explicit for loop.

|               | Apples | Beef  | Eggs | Potatoes |
| ------------- | ------ | ----- | ---- | -------- |
| carbohydrates | 56.0   | 0.0   | 4.4  | 68.0     |
| proteins      | 1.2    | 104.0 | 52.0 | 8.0      |
| fats          | 1.8    | 135.0 | 99.0 | 0.9      |

In [10]:
import numpy as np
A = np.array([
    [56.0, 0.0, 4.4, 68.0],
    [1.2, 104.0, 52.0, 8.0],
    [1.8, 135.0, 99.0, 0.9]
])
print(A)

[[ 56.    0.    4.4  68. ]
 [  1.2 104.   52.    8. ]
 [  1.8 135.   99.    0.9]]


In [11]:
cal = A.sum(axis=0)
print(cal)

[ 59.  239.  155.4  76.9]


In [14]:
percentage = A / cal.reshape(1, 4) * 100
print(percentage)

[[94.91525424  0.          2.83140283 88.42652796]
 [ 2.03389831 43.51464435 33.46203346 10.40312094]
 [ 3.05084746 56.48535565 63.70656371  1.17035111]]


This uses python broadcasting

## Explanation

When you take a vector (or a matrix), and add it to a single number, python automatically expands the single number to represent a similar shaped vector (or matrix) to add to the original vector. This is called broadcasting in python.

### Example 1:

| 1   |
| --- |
| 2   |
| 3   |
| 4   |

\+ 100 =

| 101 |
| --- |
| 102 |
| 103 |
| 104 |

### Example 2:

| 1   | 2   | 3   |
| --- | --- | --- |
| 4   | 5   | 6   |

\+

|100|200|300|
|---|---|---|

=

| 101 | 202 | 303 |
| --- | --- | --- |
| 104 | 205 | 306 |

## General Principle

If you have an (m, n) matrix and add, subtract, divide or multiply by a (1, n) matrix, then it is copied into a (m, n) matrix and vice versa for (n, m) matrix.